### PR TITLE
dyno: add input type to reduce uAST node

### PR DESCRIPTION
**Pull request by arezaii**
This PR adds the ability for uAST `Reduce` nodes to store either an
input type or a simple identifier. 

Previously, the uAST node for `reduce` did not allow for an input type
`minmax(int) reduce x` for example. This change removes the `op_`
field that stored a `UniqueString`. Instead, we store a child node
that is either an `Identifier` or a `FnCall` at index 1 in the list
of children. 

The converter for uAST to AST and the Chapel syntax printer have been
updated to account for the changes to `Reduce` node.

TESTING:

- [x] paratest
- [x] parstest with `--dyno` (115 vs. 119 failures on main)

reviewed by @dlongnecke-cray - thanks!

Signed-off-by: arezaii <ahmad.rezaii@hpe.com>

### CODE REVIEW

File: compiler/dyno/lib/parsing/ParserContextImpl.h [hunk 1]
```diff
@@ -1962,7 +1962,23 @@ AstNode* ParserContext::buildCustomReduce(YYLTYPE location,
...
                                           AstNode* lhs,
                                           AstNode* rhs) {
-  if (!lhs->isIdentifier()) {
+
+  if (lhs->isFnCall()) {
+    if (!lhs->toFnCall()->calledExpression()->isIdentifier()) {
+      const char* msg = "Expected identifier for reduction name";
+      return raiseError(locIdent, msg);
+    } else if (lhs->toFnCall()->numChildren() != 2 ) {
+      const char* msg = "for a reduce intent, the 'reduce' keyword must be preceded by the reduction operator or the name of the reduction class with the single optional argument indicating the type of the reduction input";
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/parsing/ParserContextImpl.h:1971]: Let's keep indentation at 80 cols here.

File: compiler/dyno/lib/uast/Reduce.cpp [hunk 2]
```diff
@@ -41,6 +41,22 @@ owned<Reduce> Reduce::build(Builder* builder,
   return toOwned(ret);
 }
 
+owned<Reduce> Reduce::build(Builder* builder,
+                            Location loc,
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/uast/Reduce.cpp:45]: I think we probably only need the one builder, no? Ok to just add the argument to the previous builder.

File: compiler/dyno/include/chpl/uast/Reduce.h [hunk 3]
```diff
@@ -73,6 +93,14 @@ class Reduce final : public Call {
     return op_;
   }
 
+  /**
+   Returns the input type, which may be empty.
+   An input type is specified as int in `(minmax(int) reduce sum)`
+*/
+  UniqueString inputType() const {
```
Comment (dlongnecke-cray) [inline compiler/dyno/include/chpl/uast/Reduce.h:100]: I feel like this should return `Identifier*` instead. That way we have location information.

File: compiler/dyno/lib/parsing/ParserContextImpl.h [hunk 4]
```diff
@@ -1962,7 +1962,23 @@ AstNode* ParserContext::buildCustomReduce(YYLTYPE location,
...
+      return raiseError(locIdent, msg);
+    }
+    UniqueString identifier = lhs->toFnCall()->calledExpression()->toIdentifier()->name();
```
Comment (dlongnecke-cray) [inline compiler/dyno/lib/parsing/ParserContextImpl.h:1974]: Are we sure that it can only ever be `Identifier`? What if we had a type function?

**Review by dlongnecke-cray — CHANGES_REQUESTED** [review]
Minor quibbles about column length, and then I had some questions. Mainly - I'm wondering if we should have the `inputType` return an `AstNode*`.

File: compiler/dyno/lib/uast/Reduce.cpp (hunk 2 above)
Comment (arezaii) [reply compiler/dyno/lib/uast/Reduce.cpp]: removed

File: compiler/dyno/include/chpl/uast/Reduce.h (hunk 3 above)
Comment (arezaii) [reply compiler/dyno/include/chpl/uast/Reduce.h]: Changed to store and return an `AstNode*`, which should always be a `FnCall`, and in the good case, will have one actual which is the `InputType`.

File: compiler/dyno/lib/parsing/ParserContextImpl.h (hunk 4 above)
Comment (arezaii) [reply compiler/dyno/lib/parsing/ParserContextImpl.h]: type functions aren't allowed here in the production compiler

**Review by dlongnecke-cray — APPROVED** [review]
LGTM!

**Pull request merged**